File: Daniel Lord-Doyle/week4/ruby_mta/mta.rb
lines = {
    :n => ['Times Square', '34th', '28th', '23rd', 'Union Square', '8th'],
    :l => ['8th', '6th', 'Union Square', '3rd', '1st'],
    :six => ['Grand Central', '33rd', '28th', '23rd', 'Union Square', 'Astor Place'],
}

def find_index line_symbol, stop
    lines[line_symbol].each_with_index do | stop_in_array, i |
        if stop_in_array == stop
            return i
            break
        end
    end
end

def plan_single_line_trip line_symbol, first_stop, second_stop

    first_stop_index = find_index line_symbol, first_stop
    second_stop_index = find_index line_symbol, second_stop

    trip_array = $lines[line_symbol][first_stop_index..second_stop_index]
end

def plan_trip line1, line2, stop1, stop2
    
    if line1 == line2
        if (find_index line1, stop1) < (find_index line1, stop2)
            p plan_single_line_trip line1, stop1, stop2
        else
            $lines[line1].reverse!
            p plan_single_line_trip line1, stop1, stop2
        end
        
    elsif line1 != line2

        if (find_index line1, stop1) < (find_index line1, 'Union Square')
            p plan_single_line_trip line1, stop1, 'Union Square'

            if (find_index line2, 'Union Square') < (find_index line2, stop2)
                p plan_single_line_trip line2, 'Union Square', stop2
            else (find_index line2, 'Union Square') > (find_index line2, stop2)
                $lines[line2].reverse!
                p plan_single_line_trip line2, 'Union Square', stop2
            end

        else (find_index line1, stop1) > (find_index line1, 'Union Square')
                $lines[line1].reverse!
                p plan_single_line_trip line1, stop1, 'Union Square'
                if (find_index line2, 'Union Square') < (find_index line2, stop2)
                    p plan_single_line_trip line2, 'Union Square', stop2
                else (find_index line2, 'Union Square') > (find_index line2, stop2)
                    $lines[line2].reverse!
                    p plan_single_line_trip line2, 'Union Square', stop2
                end
        end
    end
end

plan_trip :n, :n, 'Union Square', 'Times Square'
plan_trip :n, :l, '34th', '3rd'

File: Laurence Qi/week1/Day 5/mta.js
const lines = {
    N: ['Times Square', '34th', '28th', '23rd', 'Union Square', '8th'],
    L: ['8th', '6th', 'Union Square', '3rd', '1st'],
    Six: ['Grand Central', '33rd', '28th', '23rd', 'Union Square', 'Astor Place']
}
// given the following parameters, determine the route from start to the destination station print each movement, Print line changes, print remaining stations. 
const planTrip = (startLine, startStation, stopLine, stopStation) => {
    // Confirm if all parameters are inside of the objects. 
    // debugger;
    const goToDestination = (currentLine, start, stop) => {
        //Find the index of the startStation and stopStation
        let i = currentLine.indexOf(start);
        let j = currentLine.indexOf(stop);
        const increment = j - i >= currentLine.length - j + i ? 1 : -1; //determine if train goes forwards or backwards

        while (currentLine[i] !== stop) {
            console.log(currentLine[i]);
            i += increment;

            //TODO: compress into one line
            if( increment === 1 ) {
                if (i === currentLine.length) { i = 0 }; // when you go to end of array go back to beginning. 
            }
            else {
                if (i === -1) { i = currentLine.length }
            }
        }
    }
    
    let currentLine = lines[startLine];// find the position in the array where starting station is. 
    console.log(`starting at line "${startLine}" and the starting station is "${startStation}"`)

    //Station on same line. 
    if (startLine === stopLine) {
        goToDestination(currentLine, startStation, stopStation);
    }
    else{
        goToDestination(currentLine, startStation, 'Union Square');
        console.log(`Switching to line "${stopLine}"`);
        
        currentLine = lines[stopLine];
        goToDestination(currentLine, 'Union Square', stopStation);
    }

    console.log(`You have arrived at your destination "${stopStation}"`);
}

planTrip('N', '34th', 'L', '8th');
// console.log("\n Next Trip \n");
// planTrip('N', '28th', 'N', 'Times Square');
